### PR TITLE
Jb/add trace flag auto delete

### PR DESCRIPTION
**Pull request by jeffb-sfdc**
### What does this PR do?
Implements the auto removal of temporary trace flags.

### What issues does this PR fix or reference?
@W-10288383@

### Functionality Before
With the new updates to the "SFDX: Launch Apex Replay Debugger with Current File” command, a trace flag was created behind the scenes, but wasn't cleaned up after the debugging session finished.

### Functionality After
Now, after finished debugging, if the trace flag was created by the command, it is deleted from the org (existing trace flags are left untouched)

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/commands/functions/functionService.ts [hunk 1]
```diff
@@ -207,6 +208,8 @@ export class FunctionService {
         if (functionExecution) {
           functionExecution.debugSession = undefined;
         }
+
+        TraceFlagsRemover.getInstance(undefined).removeNewTraceFlags();
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-core/src/commands/functions/functionService.ts:212]: If this seems hokey, let me know.  I don't want to create another "getInstance" function, and TypeScript doesn't support function overloading.  (yes, you can sometimes simulate function overloading in TS but it's hacky and I don't care for it).

I think this is safe b/c:
1) I know that by the time we get here, `getInstance(this.connection)1 (in `createTraceFlag()`) has already been called.
2) I added guards in getInstance() to guard against this.

File: packages/salesforcedx-utils-vscode/src/helpers/traceFlagsRemover.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2022, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { Connection } from '@salesforce/core';
+
+/**
+ * TraceFlagsRemover is a singleton which deletes trace flags not created by the user.
```
Comment (jag-sfdc) [inline packages/salesforcedx-utils-vscode/src/helpers/traceFlagsRemover.ts:10]: The comment is confusing. Did you mean "deletes trace flags created by the user"?

File: packages/salesforcedx-utils-vscode/src/helpers/traceFlags.ts [hunk 3]
```diff
@@ -141,7 +142,13 @@ export class TraceFlags {
       'TraceFlag',
       traceFlag
     )) as DataRecordResult;
-    return result.success && result.id ? result.id : undefined;
+
+    if (result.success && result.id) {
+      TraceFlagsRemover.getInstance(this.connection).addNewTraceFlagId(result.id);
```
Comment (randi274) [inline packages/salesforcedx-utils-vscode/src/helpers/traceFlags.ts:147]: If the user starts a new session, would this trace flag still be removed?

File: packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlags.test.ts [hunk 4]
```diff
@@ -155,6 +155,84 @@ describe('Trace Flags', () => {
...
+      .onFirstCall()
+      .resolves({ success: true, id: 'aBcDeF' })
+      .onSecondCall()
+      .resolves({ success: true, id: '01020304' });
+
+    await traceFlags.ensureTraceFlags();
+    expect(createTraceFlagSpy.callCount).to.equal(1);
+  });
+
+  it('should verify that createTraceFlag() is not called when a trace flag exist', async () => {
```
Comment (randi274) [inline packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlags.test.ts:191]: Can we also add a test that we don't delete existing flags, and that we are able to delete the flags that we create?

Comment (jeffb-sfdc) [reply packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlags.test.ts:191]: I'm working on tests for the other files I touched.  After I finished I'll circle back to this.  I think the ask of "add a test that we don't delete existing flags" might be covered in another file (either traceFlagsRemover, or this one, traceFlags)

File: packages/salesforcedx-utils-vscode/src/helpers/traceFlags.ts (hunk 3 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-utils-vscode/src/helpers/traceFlags.ts:147]: By new session, do you if they quit VS Code and re-run?  `connection` is coming from `workspaceContext.getConnection()`.  I looked at `connection.accessToken` and it was the same value between runs, even after I quit VSC, ran VSC again, and then ran the command.

File: packages/salesforcedx-utils-vscode/src/helpers/traceFlagsRemover.ts (hunk 2 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-utils-vscode/src/helpers/traceFlagsRemover.ts:10]: done

File: packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlags.test.ts (hunk 4 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlags.test.ts:191]: I think these two tests cover both of these scenarios.  Let's sync up & discuss.

File: packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlagsRemover.test.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    const instance = TraceFlagsRemover.getInstance(mockConnection);
+    instance.addNewTraceFlagId('123');
+
+    await instance.removeNewTraceFlags();
+
+    expect(toolingDeleteStub.called).to.equal(true);
+    expect(1).to.equal(1);
+  });
+
+  it('should delete multiple trace flags', async () => {
```
[the remark this answers is not in the document]
Comment (jeffb-sfdc) [reply packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlagsRemover.test.ts:93]: done (added the, "should validate that connection.tooling.delete is not called when a trace flag already exists" test")

**Review by randi274 — APPROVED** [review]

**Review by jag-sfdc — APPROVED** [review]

**Pull request merged**